samples/nrf9160/modem_shell/src/location/location_cmd_utils.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#include <shell/shell.h>
+#include <date_time.h>
+#include <cJSON.h>
+#include <math.h>
+#include <getopt.h>
+#include <net/nrf_cloud.h>
+#include <modem/location.h>
+
+#include "mosh_print.h"
+#include "location_cmd_utils.h"
+
+/******************************************************************************/
+
+/**
+ * @brief Compute checksum for NMEA string.
+ *
+ * @param datastring - The NMEA string to compute the checksum for.
+ * @return uint8_t - The resultant checksum.
+ */
+static uint8_t location_cmd_utils_nmea_checksum(const char *const datastring)
+{
+	uint8_t checksum = 0;
+	int len = strlen(datastring);
+
+	for (int i = 0; i < len; i++) {
+		checksum ^= datastring[i];
+	}
+	return checksum;
+}
+
+int location_cmd_utils_generate_nmea(const struct location_data *location_data, char *nmea_buf,
+				 int nmea_buf_len)
+{
+	__ASSERT_NO_MSG(location_data != NULL);
+	__ASSERT_NO_MSG(nmea_buf != NULL);
+
+	/* Generate a NMEA message from the provided location lib data */
+	int latdeg = floor(fabs(location_data->latitude));
+	int londeg = floor(fabs(location_data->longitude));
+	double latmin = (fabs(location_data->latitude) - latdeg) * 60;
+	double lonmin = (fabs(location_data->longitude) - londeg) * 60;
+	int payloadlength;
+
+	snprintf(nmea_buf, nmea_buf_len, "$GPGGA,,%02d%08.5f,%c,%02d%08.5f,%c,,,,,,,,,", latdeg,
+		 latmin, location_data->latitude > 0 ? 'N' : 'S', londeg, lonmin,
+		 location_data->longitude > 0 ? 'E' : 'W');
+	payloadlength = strlen(nmea_buf);
+	snprintf(nmea_buf + payloadlength, nmea_buf_len - payloadlength, "*%02X\n",
+		 location_cmd_utils_nmea_checksum(nmea_buf + 1));
+	return payloadlength;
+}
+
+/**************************************************************************************************/
+
+#define MSG_TYPE      "messageType"
+#define MSG_APP_ID    "appId"
+#define MSG_DATA      "data"
+#define MSG_TIMESTAMP "ts"
+
+#define MSG_APP_ID_GNSS_FORMAT_NMEA "GPS"
+#define MSG_APP_ID_GNSS_FORMAT_PVT  "GNSS"
+#define MSG_TYPE_DATA               "DATA"
+
+#define MSG_TYPE_DATA_GNSS_LONGITUDE "lng"
+#define MSG_TYPE_DATA_GNSS_LATITUDE  "lat"
+#define MSG_TYPE_DATA_GNSS_ACCURACY  "acc"
+
+int location_cmd_utils_gnss_loc_to_cloud_payload_json_encode(
+	enum location_cmd_cloud_data_gnss_format format, const struct location_data *location_data,
+	char **json_str_out)
+{
+	__ASSERT_NO_MSG(location_data != NULL);
+	__ASSERT_NO_MSG(json_str_out != NULL);
+
+	/* This structure corresponds to the General Message Schema described in the
+	 * application-protocols repo:
+	 * https://github.com/nRFCloud/application-protocols
+	 */
+
+	int err = 0;
+	cJSON *root_obj = cJSON_CreateObject();
+	cJSON *gnss_data_obj = cJSON_CreateObject();
```

Comment:
This might leak memory, because it's not deleted in `cleanup`. It needs to be explicitly deleted if it has not yet been added to `root_obj`.

Actually, it seems that it leaks also every time NMEA format location is sent.

Reply:
Excellent founding, thanks!